### PR TITLE
Added support for arcee-ai/AFM-4.5B model in tt-transformers

### PR DESCRIPTION
**Pull request by ssinghalTT**
### Problem description
Add support for arcee-ai/AFM-4.5B model in tt-transformers.

HF link: https://huggingface.co/arcee-ai/AFM-4.5B

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/generator_vllm.py [hunk 1]
```diff
@@ -409,6 +409,40 @@ def allocate_kv_cache(self, *args, **kwargs):
...
 
 
+class TTArceeForCausalLM(Generator):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def initialize_vllm_model(
+        cls, hf_config, mesh_device, max_batch_size, max_seq_len=131072, n_layers=None, tt_data_parallel=1
+    ):
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator_vllm.py:426]: +1 to copilot's comment about optimizations parameter missing

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -372,7 +372,13 @@ def prefill_forward_text(
             ttnn.synchronize_device(self.model[model_id].mesh_device)
 
             # Since we give unpadded_seq_len, only the tile containing the last token is returned
-            output_logits[idx] = self.model[model_id].process_output_prefill(out, last_token_idx=(last_token_idx % 32))
+            logits_slice = self.model[model_id].process_output_prefill(out, last_token_idx=(last_token_idx % 32))
+            # Align vocab dims by slicing to common size if mismatch
+            if logits_slice.shape[-1] != output_logits.shape[-1]:
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:377]: Why does the mismatch occur?

Comment (ssinghalTT) [reply models/tt_transformers/tt/generator.py:377]: If the vocab size is not a multiple of 32.

**Review by gwangTT — COMMENTED** [review]
Thanks for contributing to tt_transformers!

The code diff to `mlp.py` is significant enough that it is probably cleaner to create a new file `mlp_2_proj.py`. The MLP module could inherit or compose the one in `mlp.py`.

In `tt_transformers/tt/decoder.py`, the new MLP module could be selected in similar ways as [this code](https://github.com/tenstorrent/tt-metal/blob/60aa5b6efd6d622649149192c8130d999eeeb9c4/models/tt_transformers/tt/decoder.py#L65-L95).

**Comment by ssinghalTT** [comment]
@gwangTT done!

File: models/tt_transformers/tt/generator_vllm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ssinghalTT) [reply models/tt_transformers/tt/generator_vllm.py:426]: Done.

**Review by dpopovTT — APPROVED** [review]

File: models/tt_transformers/tt/lm_head.py [hunk 3]
```diff
@@ -39,8 +39,13 @@ def __init__(
 
         self.model_config = args.get_model_config()
 
+        # ARCEE FIX: For arcee models with explicit padded_vocab_size, use it
+        is_arcee = "AFM" in args.model_name or "arcee" in args.model_name.lower()
```
Comment (yieldthought) [inline models/tt_transformers/tt/lm_head.py]: I don't want arcee-specific checks in here. Why does AFM need to run this codepath? Something about the vocab size dimension? If we add an extra conditional, it should be because we're making the code more general for future models, not special-casing it.

File: models/common/sampling/tt_penalties.py [hunk 4]
```diff
@@ -88,8 +88,19 @@ def __init__(self, mesh_device, args):
         num_devices = max(mesh_device.shape[-1], mesh_device.shape[-2])
         self.num_devices = num_devices
         self.needs_padding = False
-        if self.vocab_size == args.vocab_size:
-            # need to add at least one tile padding for the histogram to handle padded tokens
+
+        # ARCEE FIX: Ensure vocab_size is tile-aligned PER DEVICE for arcee models
+        # Even if padded_vocab_size is set, it might not be tile-aligned per device after sharding
+        is_arcee = "AFM" in args.model_name or "arcee" in args.model_name.lower()
```
Comment (yieldthought) [inline models/common/sampling/tt_penalties.py]: Is this about vocab size not being a multiple of 32? In that case I'm happy to see us adding support for non-tile-aligned vocab sizes, but it should be gated behind the vocab size not being tile aligned, not behind a specific model name.

**Review by yieldthought — CHANGES_REQUESTED** [review]
I appreciate the work done to extend TTT to support models with 2-proj MLPs and would really like to see whatever's going on with the vocab treated in the same way and not special-cased to this model name.

**Comment by yieldthought** [comment]
Can you please generate reference logits and run the demo with teacher forcing mode to collect and fill PERF.md with performance and accuracy expectations?

File: models/tt_transformers/tt/mlp_2_proj.py
Comment (yieldthought) [inline models/tt_transformers/tt/mlp_2_proj.py:1]: This needs to be covered by tests/test_mlp.py or by its own unit test.

**Review by yieldthought — CHANGES_REQUESTED** [review]

**Comment by yieldthought** [comment]
Codex review finding #1: In ModelArgs.load_state_dict, vocab_size can be corrected from HF weights but padded_vocab_size is not recomputed, so arcee can end up with padded_vocab_size < vocab_size; LMHead then pads/ shards with the smaller size. Consider recomputing padded_vocab_size after vocab_size is updated. (Codex)

**Comment by yieldthought** [comment]
Codex review finding #2: In MLP2Proj.forward on Galaxy, activation is applied before reduce_scatter/all_reduce and the reduced w3_out is discarded; w2_in uses the unreduced activation. This looks incorrect for TG sharded paths. (Codex)

**Comment by yieldthought** [comment]
Codex review finding #3: hidden_act=relu2 maps to the string "relu2" in ModelArgs, but MLP (3-proj) passes activation into ttnn.mul as an op type; any 3-proj model using relu2 likely fails. Consider adding an explicit relu2 branch or mapping to a supported unary op. (Codex)

**Comment by yieldthought** [comment]
Please address or refute all three of the codex review comments, they look valid to me.

**Comment by ssinghalTT** [comment]
@yieldthought done with the changes.

File: models/tt_transformers/tt/generator_vllm.py [hunk 5]
```diff
@@ -409,6 +409,47 @@ def allocate_kv_cache(self, *args, **kwargs):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def initialize_vllm_model(
+        cls,
+        hf_config,
+        mesh_device,
+        max_batch_size,
+        max_seq_len=131072,
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator_vllm.py:422]: please remove the default value for max_seq_len

File: models/tt_transformers/tt/generator_vllm.py [hunk 6]
```diff
@@ -409,6 +409,47 @@ def allocate_kv_cache(self, *args, **kwargs):
...
+    ):
+        tt_model, model_args = initialize_vllm_text_transformer(
+            hf_config,
+            tt_data_parallel,
+            mesh_device,
+            max_batch_size,
+            max_seq_len=max_seq_len,
+            n_layers=n_layers,
+            dtype=ttnn.bfloat8_b,
+            optimizations=DecodersPrecision.from_string(optimizations),
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator_vllm.py:435]: Should check `if optimizations is not None` similar to other model classes

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py:377]: > If the vocab size is not a multiple of 32.

Could you modify the condition to check for that explicitly instead of doing != ?

**Review by yieldthought — APPROVED** [review]
🥇 Thank-you!
